# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    
    source_id_text = [[source_vocab_to_int.setdefault(word,source_vocab_to_int['<UNK>']) for word in line.split()] for line in source_text.split('\n')]
    target_id_text = [[target_vocab_to_int.setdefault(word,source_vocab_to_int['<UNK>']) for word in line.split()] + [target_vocab_to_int['<EOS>']] for line in target_text.split('\n')] 
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input_data = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='targets')
    learning_rate=tf.placeholder(tf.float32,name='learning_rate')
    keep_prob=tf.placeholder(tf.float32,name='keep_prob')
    
    target_seq_length=tf.placeholder(tf.int32,(None,),name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_seq_length, name='max_target_len')
    source_seq_length=tf.placeholder(tf.int32,(None,),name='source_sequence_length')
    return input_data, targets, learning_rate, keep_prob, target_seq_length, max_target_sequence_length,source_seq_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    #Implement process_decoder_input by removing the last word id from each batch in target_data 
    #and concat the GO ID to the begining of each batch.
    dec_i = tf.strided_slice(target_data,[0, 0], [batch_size,-1], [1, 1])
    return tf.concat([tf.fill([batch_size,1],target_vocab_to_int['<GO>']),dec_i],1)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    #Embed the encoder input using tf.contrib.layers.embed_sequence
    embed_input= tf.contrib.layers.embed_sequence(rnn_inputs,source_vocab_size,encoding_embedding_size)
    
    #Construct a stacked tf.contrib.rnn.LSTMCell wrapped in a tf.contrib.rnn.DropoutWrapper
    lstmcell =[tf.contrib.rnn.LSTMCell(rnn_size,
                        initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2)) for _ in range(num_layers)]
    cell=tf.contrib.rnn.MultiRNNCell(lstmcell)
    drop = tf.contrib.rnn.DropoutWrapper(cell, output_keep_prob=keep_prob)
    
    #Pass cell and embedded input to tf.nn.dynamic_rnn()
    enc_output, enc_state = tf.nn.dynamic_rnn(drop, embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    # Create a tf.contrib.seq2seq.TrainingHelper
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                            sequence_length=target_sequence_length,
                                                            time_major=False)
        
        
    # Create a tf.contrib.seq2seq.BasicDecoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                           training_helper,
                                                           encoder_state,
                                                           output_layer) 
        
    # Obtain the decoder outputs from tf.contrib.seq2seq.dynamic_decode
    #1.1
    training_decoder_output,_= tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                       impute_finished=True,
                                                                       maximum_iterations=max_summary_length)
    #1.2
    #training_decoder_output,a,b= tf.contrib.seq2seq.dynamic_decode(training_decoder,
     #                                                                  impute_finished=True,
      #                                                                 maximum_iterations=max_summary_length)
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    #Create a tf.contrib.seq2seq.GreedyEmbeddingHelper

    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                                start_tokens,
                                                                end_of_sequence_id)

    #Create a tf.contrib.seq2seq.BasicDecoder 
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                        inference_helper,
                                                        encoder_state,
                                                        output_layer)
        
    # Obtain the decoder outputs from tf.contrib.seq2seq.dynamic_decode
    #1.1
    inference_decoder_output,_= tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                            impute_finished=True,
                                                            maximum_iterations=max_target_sequence_length)
    
    #1.2
    #inference_decoder_output,a,b= tf.contrib.seq2seq.dynamic_decode(inference_decoder,
     #                                                       impute_finished=True,
      #                                                      maximum_iterations=max_target_sequence_length)
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # Embed the target sequences
    with tf.variable_scope('decoding_layer') as decoding_layer:
        #Modified as requested
        dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
        dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
        #dec_embed_input= tf.contrib.layers.embed_sequence(dec_input,target_vocab_size,decoding_embedding_size)
        #Construct the decoder LSTM cell (just like you constructed the encoder cell above)

        lstmcell =[tf.contrib.rnn.LSTMCell(rnn_size,
                        initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2)) for _ in range(num_layers)]
        
        cell=tf.contrib.rnn.MultiRNNCell(lstmcell)
        # Create an output layer to map the outputs of the decoder to the elements of our vocabulary

        output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    # Use the your decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob) function to get the training logits.

    with tf.variable_scope('decoding_layer') as decoding_layer:
        train_logits=decoding_layer_train(encoder_state,cell,dec_embed_input,target_sequence_length,max_target_sequence_length,output_layer,keep_prob)
    
    # Use your decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob) function to get the inference logits
    with tf.variable_scope("decoding_layer", reuse=True):
        inference_logits=decoding_layer_infer(encoder_state,cell,dec_embeddings,target_vocab_to_int['<GO>'],target_vocab_to_int['<EOS>']
                                          ,max_target_sequence_length-1,target_vocab_size,output_layer,batch_size,keep_prob)

    return train_logits, inference_logits



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # Encode the input using your encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size).
    enc_output, enc_state=encoding_layer(input_data,rnn_size,num_layers,keep_prob,source_sequence_length,source_vocab_size,enc_embedding_size)
    #Process target data using your process_decoder_input(target_data, target_vocab_to_int, batch_size) function.
    dec_input=process_decoder_input(target_data,target_vocab_to_int,batch_size)
    #Decode the encoded input using your decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size) function.
    training_decoder_output, inference_decoder_output=decoding_layer(dec_input,enc_state,target_sequence_length,
                                                                     max_target_sentence_length,rnn_size,num_layers,
                                                                     target_vocab_to_int,target_vocab_size, batch_size,keep_prob, dec_embedding_size)
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 12
# Batch Size
batch_size = 512
# RNN Size
rnn_size = 512
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 12
decoding_embedding_size = 12
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.5
display_step = 5

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})

            if batch_i % display_step == 0 and batch_i > 0:

                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                    # targets:target_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})

                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     #targets:valid_targets_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    5/269 - Train Accuracy: 0.2668, Validation Accuracy: 0.3416, Loss: 4.2144
Epoch   0 Batch   10/269 - Train Accuracy: 0.2764, Validation Accuracy: 0.3469, Loss: 3.3696
Epoch   0 Batch   15/269 - Train Accuracy: 0.3318, Validation Accuracy: 0.3729, Loss: 3.0932
Epoch   0 Batch   20/269 - Train Accuracy: 0.3118, Validation Accuracy: 0.3766, Loss: 3.0510
Epoch   0 Batch   25/269 - Train Accuracy: 0.3189, Validation Accuracy: 0.3805, Loss: 3.0180
Epoch   0 Batch   30/269 - Train Accuracy: 0.3605, Validation Accuracy: 0.3932, Loss: 2.8414
Epoch   0 Batch   35/269 - Train Accuracy: 0.3634, Validation Accuracy: 0.3908, Loss: 2.7557
Epoch   0 Batch   40/269 - Train Accuracy: 0.3339, Validation Accuracy: 0.3937, Loss: 2.8448
Epoch   0 Batch   45/269 - Train Accuracy: 0.3264, Validation Accuracy: 0.3896, Loss: 2.8222
Epoch   0 Batch   50/269 - Train Accuracy: 0.3274, Validation Accuracy: 0.3825, Loss: 2.8512
Epoch   0 Batch   55/269 - Train Accuracy: 0.3667, Validation Accuracy

Epoch   1 Batch  185/269 - Train Accuracy: 0.5460, Validation Accuracy: 0.5537, Loss: 0.6221
Epoch   1 Batch  190/269 - Train Accuracy: 0.5211, Validation Accuracy: 0.5400, Loss: 0.6127
Epoch   1 Batch  195/269 - Train Accuracy: 0.5499, Validation Accuracy: 0.5516, Loss: 0.6232
Epoch   1 Batch  200/269 - Train Accuracy: 0.5464, Validation Accuracy: 0.5694, Loss: 0.6382
Epoch   1 Batch  205/269 - Train Accuracy: 0.5172, Validation Accuracy: 0.5337, Loss: 0.6091
Epoch   1 Batch  210/269 - Train Accuracy: 0.5641, Validation Accuracy: 0.5756, Loss: 0.6066
Epoch   1 Batch  215/269 - Train Accuracy: 0.5883, Validation Accuracy: 0.5610, Loss: 0.5835
Epoch   1 Batch  220/269 - Train Accuracy: 0.5654, Validation Accuracy: 0.5558, Loss: 0.5803
Epoch   1 Batch  225/269 - Train Accuracy: 0.5505, Validation Accuracy: 0.5711, Loss: 0.6160
Epoch   1 Batch  230/269 - Train Accuracy: 0.5418, Validation Accuracy: 0.5566, Loss: 0.6167
Epoch   1 Batch  235/269 - Train Accuracy: 0.5651, Validation Accuracy

Epoch   3 Batch  100/269 - Train Accuracy: 0.6669, Validation Accuracy: 0.6468, Loss: 0.4517
Epoch   3 Batch  105/269 - Train Accuracy: 0.6390, Validation Accuracy: 0.6323, Loss: 0.4674
Epoch   3 Batch  110/269 - Train Accuracy: 0.6137, Validation Accuracy: 0.6180, Loss: 0.4469
Epoch   3 Batch  115/269 - Train Accuracy: 0.6229, Validation Accuracy: 0.6510, Loss: 0.4875
Epoch   3 Batch  120/269 - Train Accuracy: 0.6354, Validation Accuracy: 0.6562, Loss: 0.4762
Epoch   3 Batch  125/269 - Train Accuracy: 0.6524, Validation Accuracy: 0.6577, Loss: 0.4576
Epoch   3 Batch  130/269 - Train Accuracy: 0.6205, Validation Accuracy: 0.6491, Loss: 0.4681
Epoch   3 Batch  135/269 - Train Accuracy: 0.6272, Validation Accuracy: 0.6578, Loss: 0.4827
Epoch   3 Batch  140/269 - Train Accuracy: 0.6524, Validation Accuracy: 0.6428, Loss: 0.4568
Epoch   3 Batch  145/269 - Train Accuracy: 0.6390, Validation Accuracy: 0.6413, Loss: 0.4493
Epoch   3 Batch  150/269 - Train Accuracy: 0.6630, Validation Accuracy

Epoch   5 Batch   15/269 - Train Accuracy: 0.7140, Validation Accuracy: 0.7176, Loss: 0.3195
Epoch   5 Batch   20/269 - Train Accuracy: 0.6955, Validation Accuracy: 0.7136, Loss: 0.3357
Epoch   5 Batch   25/269 - Train Accuracy: 0.6968, Validation Accuracy: 0.7281, Loss: 0.3337
Epoch   5 Batch   30/269 - Train Accuracy: 0.7301, Validation Accuracy: 0.7303, Loss: 0.3101
Epoch   5 Batch   35/269 - Train Accuracy: 0.7186, Validation Accuracy: 0.7309, Loss: 0.3247
Epoch   5 Batch   40/269 - Train Accuracy: 0.7031, Validation Accuracy: 0.7262, Loss: 0.3351
Epoch   5 Batch   45/269 - Train Accuracy: 0.7033, Validation Accuracy: 0.7234, Loss: 0.3207
Epoch   5 Batch   50/269 - Train Accuracy: 0.6888, Validation Accuracy: 0.7138, Loss: 0.3267
Epoch   5 Batch   55/269 - Train Accuracy: 0.7499, Validation Accuracy: 0.7300, Loss: 0.3024
Epoch   5 Batch   60/269 - Train Accuracy: 0.7366, Validation Accuracy: 0.7235, Loss: 0.2972
Epoch   5 Batch   65/269 - Train Accuracy: 0.7353, Validation Accuracy

Epoch   6 Batch  195/269 - Train Accuracy: 0.8170, Validation Accuracy: 0.8020, Loss: 0.1800
Epoch   6 Batch  200/269 - Train Accuracy: 0.8050, Validation Accuracy: 0.8041, Loss: 0.1791
Epoch   6 Batch  205/269 - Train Accuracy: 0.8028, Validation Accuracy: 0.7997, Loss: 0.1761
Epoch   6 Batch  210/269 - Train Accuracy: 0.8194, Validation Accuracy: 0.8126, Loss: 0.1763
Epoch   6 Batch  215/269 - Train Accuracy: 0.8282, Validation Accuracy: 0.8249, Loss: 0.1635
Epoch   6 Batch  220/269 - Train Accuracy: 0.8274, Validation Accuracy: 0.8109, Loss: 0.1634
Epoch   6 Batch  225/269 - Train Accuracy: 0.8251, Validation Accuracy: 0.8216, Loss: 0.1647
Epoch   6 Batch  230/269 - Train Accuracy: 0.8392, Validation Accuracy: 0.8334, Loss: 0.1722
Epoch   6 Batch  235/269 - Train Accuracy: 0.8475, Validation Accuracy: 0.8229, Loss: 0.1576
Epoch   6 Batch  240/269 - Train Accuracy: 0.8451, Validation Accuracy: 0.8299, Loss: 0.1479
Epoch   6 Batch  245/269 - Train Accuracy: 0.8265, Validation Accuracy

Epoch   8 Batch  110/269 - Train Accuracy: 0.8821, Validation Accuracy: 0.8603, Loss: 0.0979
Epoch   8 Batch  115/269 - Train Accuracy: 0.8766, Validation Accuracy: 0.8716, Loss: 0.1030
Epoch   8 Batch  120/269 - Train Accuracy: 0.8902, Validation Accuracy: 0.8810, Loss: 0.1007
Epoch   8 Batch  125/269 - Train Accuracy: 0.8910, Validation Accuracy: 0.8674, Loss: 0.0994
Epoch   8 Batch  130/269 - Train Accuracy: 0.8904, Validation Accuracy: 0.8828, Loss: 0.0969
Epoch   8 Batch  135/269 - Train Accuracy: 0.8716, Validation Accuracy: 0.8803, Loss: 0.1039
Epoch   8 Batch  140/269 - Train Accuracy: 0.8948, Validation Accuracy: 0.8819, Loss: 0.0984
Epoch   8 Batch  145/269 - Train Accuracy: 0.9012, Validation Accuracy: 0.8825, Loss: 0.0876
Epoch   8 Batch  150/269 - Train Accuracy: 0.8812, Validation Accuracy: 0.8941, Loss: 0.0980
Epoch   8 Batch  155/269 - Train Accuracy: 0.8888, Validation Accuracy: 0.8888, Loss: 0.0846
Epoch   8 Batch  160/269 - Train Accuracy: 0.8954, Validation Accuracy

Epoch  10 Batch   25/269 - Train Accuracy: 0.8969, Validation Accuracy: 0.8972, Loss: 0.0763
Epoch  10 Batch   30/269 - Train Accuracy: 0.9102, Validation Accuracy: 0.9002, Loss: 0.0644
Epoch  10 Batch   35/269 - Train Accuracy: 0.9214, Validation Accuracy: 0.9064, Loss: 0.0744
Epoch  10 Batch   40/269 - Train Accuracy: 0.9143, Validation Accuracy: 0.9144, Loss: 0.0637
Epoch  10 Batch   45/269 - Train Accuracy: 0.9212, Validation Accuracy: 0.9123, Loss: 0.0594
Epoch  10 Batch   50/269 - Train Accuracy: 0.9046, Validation Accuracy: 0.9083, Loss: 0.0650
Epoch  10 Batch   55/269 - Train Accuracy: 0.9256, Validation Accuracy: 0.9119, Loss: 0.0578
Epoch  10 Batch   60/269 - Train Accuracy: 0.9304, Validation Accuracy: 0.9158, Loss: 0.0549
Epoch  10 Batch   65/269 - Train Accuracy: 0.9202, Validation Accuracy: 0.9102, Loss: 0.0545
Epoch  10 Batch   70/269 - Train Accuracy: 0.9280, Validation Accuracy: 0.9055, Loss: 0.0535
Epoch  10 Batch   75/269 - Train Accuracy: 0.9312, Validation Accuracy

Epoch  11 Batch  205/269 - Train Accuracy: 0.9144, Validation Accuracy: 0.9081, Loss: 0.0420
Epoch  11 Batch  210/269 - Train Accuracy: 0.9096, Validation Accuracy: 0.9170, Loss: 0.0459
Epoch  11 Batch  215/269 - Train Accuracy: 0.9205, Validation Accuracy: 0.9231, Loss: 0.0449
Epoch  11 Batch  220/269 - Train Accuracy: 0.9292, Validation Accuracy: 0.9288, Loss: 0.0437
Epoch  11 Batch  225/269 - Train Accuracy: 0.9226, Validation Accuracy: 0.9269, Loss: 0.0414
Epoch  11 Batch  230/269 - Train Accuracy: 0.9259, Validation Accuracy: 0.9221, Loss: 0.0398
Epoch  11 Batch  235/269 - Train Accuracy: 0.9513, Validation Accuracy: 0.9120, Loss: 0.0364
Epoch  11 Batch  240/269 - Train Accuracy: 0.9299, Validation Accuracy: 0.9227, Loss: 0.0380
Epoch  11 Batch  245/269 - Train Accuracy: 0.9147, Validation Accuracy: 0.9267, Loss: 0.0477
Epoch  11 Batch  250/269 - Train Accuracy: 0.9280, Validation Accuracy: 0.9172, Loss: 0.0449
Epoch  11 Batch  255/269 - Train Accuracy: 0.9158, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [20]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    # Convert the sentence to lowercase
    sent=sentence.lower()
    # Convert words into ids using vocab_to_int
    return  [vocab_to_int.get(word,vocab_to_int['<UNK>']) for word in sent.split()]



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [21]:
#This piece of code is not working. 
#Really appreciate any feedback on this, I know the error is because of targets not being specified in feed dict but i couldnt 
# really figure out the right target value to be provided

translate_sentence = 'he saw a old yellow truck .'

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)


loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [123, 102, 46, 76, 91, 34, 8]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [36, 290, 65, 85, 36, 35, 95, 1]
  French Words: son fruit préféré est son favori . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.